File: api_nestjs_distributed_order_platform/src/application/use-case/product/create-product.use-case.ts
import { ConflictException, Inject, Injectable } from "@nestjs/common";
import type { ProductRepositoryInterface } from "src/application/port/product.repository";
import type { StockRepositoryInterface } from "src/application/port/stock.repository";
import { ProductEntity } from "src/domain/entities/product.entity";
import { StockEntity } from "src/domain/entities/stock.entity";
import { CreateProductInput } from "src/shared/interfaces/product.interface";
import { PRODUCT_REPOSITORY } from "src/shared/tokens_nest/product.token";
import { STOCK_REPOSITORY } from "src/shared/tokens_nest/stock.token";

@Injectable()
export class CreateProductUseCase {
    constructor(
        @Inject(PRODUCT_REPOSITORY)
        private readonly productRepository: ProductRepositoryInterface,
        @Inject(STOCK_REPOSITORY)
        private readonly stockRepository: StockRepositoryInterface,
    ) {}

    async execute(input: CreateProductInput): Promise<ProductEntity> {
        const existing = await this.productRepository.findBySku(input.sku);
        if (existing) {
            throw new ConflictException('Product with this SKU already exists');
        }

        const product = new ProductEntity(
            input.sku,
            input.name,
            input.price,
            input.status,
        );

        const created = await this.productRepository.create(product);

        await this.stockRepository.create(
            new StockEntity(created.getId(), input.initialStock ?? 0, 0),
        );

        return created;
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/application/usecase/product/CreateProductUseCase.java
package com.project.order.application.usecase.product;

import com.project.order.application.exception.ConflictException;
import com.project.order.application.port.ProductRepository;
import com.project.order.application.port.StockRepository;
import com.project.order.domain.model.Product;
import com.project.order.domain.model.Stock;
import org.springframework.stereotype.Service;

@Service
public class CreateProductUseCase {

    private final ProductRepository productRepository;
    private final StockRepository stockRepository;

    public CreateProductUseCase(ProductRepository productRepository, StockRepository stockRepository) {
        this.productRepository = productRepository;
        this.stockRepository = stockRepository;
    }

    public Product execute(String sku, String name, double price, int initialStock) {
        if (productRepository.findBySku(sku).isPresent()) {
            throw new ConflictException("Product with this SKU already exists");
        }

        Product created = productRepository.create(Product.create(sku, name, price));

        stockRepository.create(Stock.create(created.getId(), initialStock, 0));

        return created;
    }
}
